flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/FloorCeilCallGen.scala

```diff
@@ -105,13 +105,27 @@ class FloorCeilCallGen(
             case _ =>
               operand.resultType.getTypeRoot match {
                 case LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE =>
-                  val longTerm = s"${terms.head}.getMillisecond()"
-                  s"""
-                     |$TIMESTAMP_DATA.fromEpochMillis(
-                     |  ${qualifyMethod(arithmeticIntegralMethod.get)}(
-                     |    $longTerm,
-                     |    (long) ${unit.startUnit.multiplier.intValue()}))
+                  val millis = s"${terms.head}.getMillisecond()"
+
+                  unit match {
+                    case MILLISECOND =>
+                      val nanos =
+                        s"${qualifyMethod(arithmeticIntegralMethod.get)}(${terms.head}.getNanoOfMillisecond(), " +
```

Comment:
I think we can simplify this, we can just get the nanos here, so all 9 digits, no?

Reply:
Not sure if I understand what you mean, however I simplified that with
```scala
s"""
   | $TIMESTAMP_DATA.fromEpochMillis(
   |    $millis + ${qualifyMethod(arithmeticIntegralMethod.get)}(${terms.head}.getNanoOfMillisecond(), 1000_000L) / 1000_000L)
"""stripMargin
```

Reply:
Can't we use directly `nanos of second` instead of separately `millis` and then `nanos of millisecond`?

Reply:
The problem is that `org.apache.flink.table.data.TimestampData` contains these `millis`, `nanoOfMillis` in 2 separate variables https://github.com/apache/flink/blob/90e98ba7c858b740365fc9ffdf0afc7d791373f2/flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/TimestampData.java#L46-L50 .
It means that to get `nanos of second` it is required to get both `nano of millisecond` and `millis` and then convert to `nanos of second` since there is no corresponding methods.

Or another way is extending `org.apache.flink.table.data.TimestampData` to provide getter of `nanos of second` and probably the ability of being initialized by `nanos of second`

Reply:
I see, thx for explaining!!

Reply:
It's fine to leave it as is!